pytools/persistent_dict.py

```diff
@@ -549,112 +447,122 @@ def __init__(self, identifier: str,
-                    cache_dir,
-                    "pdict-v4-{}-py{}".format(
-                        identifier,
-                        ".".join(str(i) for i in sys.version_info)))
+        self.filename = join(container_dir, f"pdict-v5-{identifier}"
+                             + ".".join(str(i) for i in sys.version_info)
+                             + ".sqlite")
 
         self.container_dir = container_dir
-
         self._make_container_dir()
 
-    @staticmethod
-    def _warn(msg: str, category: Any = UserWarning, stacklevel: int = 0) -> None:
-        from warnings import warn
-        warn(msg, category, stacklevel=1 + stacklevel)
-
-    def store_if_not_present(self, key: K, value: V,
-                             _stacklevel: int = 0) -> None:
-        """Store (*key*, *value*) if *key* is not already present."""
-        self.store(key, value, _skip_if_present=True, _stacklevel=1 + _stacklevel)
-
-    def store(self, key: K, value: V, _skip_if_present: bool = False,
-              _stacklevel: int = 0) -> None:
-        """Store (*key*, *value*) in the dictionary."""
-        raise NotImplementedError()
-
-    def fetch(self, key: K, _stacklevel: int = 0) -> V:
-        """Return the value associated with *key* in the dictionary."""
-        raise NotImplementedError()
-
-    @staticmethod
-    def _read(path: str) -> V:
-        from pickle import load
-        with open(path, "rb") as inf:
-            return load(inf)
-
-    @staticmethod
-    def _write(path: str, value: V) -> None:
-        from pickle import HIGHEST_PROTOCOL, dump
-        with open(path, "wb") as outf:
-            dump(value, outf, protocol=HIGHEST_PROTOCOL)
-
-    def _item_dir(self, hexdigest_key: str) -> str:
-        from os.path import join
+        # isolation_level=None: enable autocommit mode
+        # https://www.sqlite.org/lang_transaction.html#implicit_versus_explicit_transactions
+        self.conn = sqlite3.connect(self.filename, isolation_level=None)
 
-        # Some file systems limit the number of directories in a directory.
-        # For ext4, that limit appears to be 64K for example.
-        # This doesn't solve that problem, but it makes it much less likely
+        self.conn.execute(
+            "CREATE TABLE IF NOT EXISTS dict "
+            "(keyhash TEXT NOT NULL PRIMARY KEY, key_value TEXT NOT NULL)"
+            )
 
-        return join(self.container_dir,
-                hexdigest_key[:3],
-                hexdigest_key[3:6],
-                hexdigest_key[6:])
+        # https://www.sqlite.org/wal.html
+        if enable_wal:
+            self.conn.execute("PRAGMA journal_mode = 'WAL'")
 
-    def _key_file(self, hexdigest_key: str) -> str:
-        from os.path import join
-        return join(self._item_dir(hexdigest_key), "key")
+        # temp_store=2: use in-memory temp store
+        # https://www.sqlite.org/pragma.html#pragma_temp_store
+        self.conn.execute("PRAGMA temp_store = 2")
```

Comment:
Could you roughly describe the findings that led to each of these?

Reply:
I added a comment regarding this in 51666d0.